python/lsst/cp/pipe/ptc.py

```diff
@@ -367,11 +434,269 @@ def runDataRef(self, dataRef, visitPairs):
                 except OperationalError:
                     pass
 
+        if self.config.doCovariancesAstier:
+            # Calculate the full covariances as in Astier+19. The variances for the PTC will be
+            # cov[0,0].
+
+            tupleCovariancesWithTags = self.computeCovariancesAstier(dataRef, visitPairs, detector)
+            # use the np.recarray to obtain the CovFit objects
+            covFits, covFitsNoB = fitData(tupleCovariancesWithTags, maxMu=self.config.maxMeanSignal,
+                                          maxMuElectrons=self.config.maxMeanSignal,
+                                          r=self.config.maximumRangeCovariancesAstier)
+
+            if self.config.makePlots:
+                self.plotCovariancesAstier(dataRef, covFits, covFitsNoB, tupleCovariancesWithTags,
+                                           log=self.log)
+
+            datasetCovAstier = PhotonTransferCurveCovAstierDataset()
+            datasetCovAstier.covariancesTuple = tupleCovariancesWithTags
+            datasetCovAstier.covariancesFits = covFits
+            datasetCovAstier.covariancesFitsWithNoB = covFitsNoB
+
+            dataset = self.getOutputPtcDataCovAstier(covFits)
+
+            datasetCovAstier.gain = dataset.gain
+            datasetCovAstier.noise = dataset.noise
+            datasetCovAstier.meanSignal = dataset.signal
+            datasetCovAstier.var = dataset.var
+            datasetCovAstier.varModel = dataset.varModel
+            datasetCovAstier.meanSignalAdu = dataset.signalAdu
+            datasetCovAstier.varAdu = dataset.varAdu
+            datasetCovAstier.varModelAdu = dataset.varModelAdu
+            datasetCovAstier.a00 = dataset.a00
+            datasetCovAstier.b00 = dataset.b00
+
+            self.log.info(f"Writing Astier+19 covariances data")
+            dataRef.put(datasetCovAstier, datasetType="photonTransferCurveCovAstierDataset")
```

Comment:
Can this method be transformed into a runDataRef()->run() chain, where runDataRef loads all necessary data, passes it to run, which generates the outputs, returns them in a Struct to runDataRef, which writes the outputs?

Reply:
I tried to clean this up a bit (e.g., disentangling the non-linearity function, trying to do data handling and I/O in dataRef, etc.).